### PR TITLE
Collect all the training_step outputs for training epoch end

### LINKED ISSUE #2320: Partially missing training_step outputs in training_epoch_end
**Learning Model:** the model consists of two branches: **teacher and student**. Two losses have been used to train two branches of the learning model using two optimizers.

**ISSUE:** In the **`training_step`**, I produced two set of metrics outputs for teacher (_optimizer_idx=0_) and student (_optimizer_idx=1_) branches (teacher: `{loss,log:{acc, f1}}` and student: `{loss,log:{acc,f1,precision}}` ). But in the **`training_epoch_end`**, I only got the combined outputs for only student outputs (_optimizer_idx=1_). All the teacher outputs (_optimizer_idx=0_) are missing.

I also see the training loop of PL and didn't observe any issue when the training loop tries to combine the **`training_step`** outputs. I am not sure what am I missing? 

 - Version: 0.8.2.dev0

### PR DESCRIPTION
**Pull request by mmiakashs**
Fixes #2320 
Bug Fixes of #2320 after the #2328 commits

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -467,7 +467,7 @@ def run_training_epoch(self):
             # only track outputs when user implements training_epoch_end
             # otherwise we will build up unnecessary memory
             if self.is_overridden('training_epoch_end', model=self.get_model()):
-                epoch_output.append(batch_output.training_step_output_for_epoch_end)
+                epoch_output += batch_output.training_step_output_for_epoch_end
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:470]: this is equivalent... why do we need to change it?

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -596,6 +596,9 @@ def run_training_batch(self, batch, batch_idx):
         # track metrics to log
         batch_log_metrics = []
 
+        # track all training_step output for epoch_end
+        all_training_step_output_for_epoch_end = []
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:600]: this is not needed... we're already tracking it at the top of the file

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -690,7 +694,7 @@ def run_training_batch(self, batch, batch_idx):
             signal=0,
             grad_norm_dic=grad_norm_dic,
             batch_log_metrics=batch_log_metrics,
-            training_step_output_for_epoch_end=opt_closure_result.training_step_output_for_epoch_end
+            training_step_output_for_epoch_end=all_training_step_output_for_epoch_end
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:697]: i believe the function that calls this method also needs to be fixed to take into account the list of outputs.

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -690,7 +694,7 @@ def run_training_batch(self, batch, batch_idx):
             signal=0,
             grad_norm_dic=grad_norm_dic,
             batch_log_metrics=batch_log_metrics,
-            training_step_output_for_epoch_end=opt_closure_result.training_step_output_for_epoch_end
+            training_step_output_for_epoch_end=training_step_output_for_epoch_end_list
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/training_loop.py:697]: this here is now a list of outputs. This means we need to make sure the method which called this processes that.

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (mmiakashs) [reply pytorch_lightning/trainer/training_loop.py:470]: It will create a list of list dict outputs(i.e. [[{},{}], [{},{}]]). However, += will create only the list of dict outputs

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (mmiakashs) [reply pytorch_lightning/trainer/training_loop.py]: `training_step_output_for_epoch_end=opt_closure_result.training_step_output_for_epoch_end`

this line only capture the last iteration outputs and discard the previous iterations.

File: pytorch_lightning/trainer/training_loop.py (hunk 4 above)
Comment (mmiakashs) [reply pytorch_lightning/trainer/training_loop.py:697]: sounds good :smile:

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py:470]: Will, it is not equal as for the original it adds element the new merge lists... Equivalent would be
```
a = [1]
b = 2
a.append(b)
b += [b]
```

**Comment by williamFalcon** [comment]
wait... i was thinking about this. I still think this PR is not correct.

We don't want the output from EVERY tbptt split... only the last one which is what this code does today

**Comment by mmiakashs** [comment]
> wait... i was thinking about this. I still think this PR is not correct.
> 
> We don't want the output from EVERY tbptt split... only the last one which is what this code does today

In that case, we will miss some outputs, which will lead to the wrong metric calculation at the end of the epoch. isn't it?

**Comment by Borda** [comment]
How is it going here?

**Comment by mmiakashs** [comment]
> How is it going here?

I couldn't able to figure out the failed tests issue. However, I am using this PR changes in my local machine and it is working perfectly.

**Comment by mmiakashs** [comment]
@Borda and @williamFalcon guys did you get the chance to look into this issue? I was using this PR version and it is working properly at my end. I didn't find the cause of this conflict.

**Comment by Borda** [comment]
is there still the issue, as #2320 is closed?

**Comment by williamFalcon** [comment]
Finished in #2890
